I am still seeing failures that look related to precision issues, but when I enable `float32` I'm seeing issues related to GPU memory:
```
[2024-06-24T22:34:08Z] =========================== short test summary info ============================
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_e2e_greedy_correctness[1-1-256-test_llm_kwargs0-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: No available memory for the cache blocks. Try increasing `gpu_memory_utilization` when initializing the engine.
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_e2e_greedy_correctness[1-32-256-test_llm_kwargs0-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: Memory of devices devices=[0] not free after dur_s=60.06 (threshold_bytes/2**30=2.0)
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_different_k[1-32-2-test_llm_kwargs0-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: No available memory for the cache blocks. Try increasing `gpu_memory_utilization` when initializing the engine.
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_different_k[1-32-2-test_llm_kwargs1-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: Memory of devices devices=[0] not free after dur_s=60.07 (threshold_bytes/2**30=2.0)
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_different_k[1-32-2-test_llm_kwargs2-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: No available memory for the cache blocks. Try increasing `gpu_memory_utilization` when initializing the engine.
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_different_k[1-32-2-test_llm_kwargs3-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: Memory of devices devices=[0] not free after dur_s=60.06 (threshold_bytes/2**30=2.0)
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_different_k[1-32-2-test_llm_kwargs4-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: No available memory for the cache blocks. Try increasing `gpu_memory_utilization` when initializing the engine.
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_disable_queue[1-32-1-test_llm_kwargs0-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: Memory of devices devices=[0] not free after dur_s=60.06 (threshold_bytes/2**30=2.0)
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_mlp_correctness.py::test_mlp_disable_queue[1-32-5-test_llm_kwargs0-baseline_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: No available memory for the cache blocks. Try increasing `gpu_memory_utilization` when initializing the engine.
  | [2024-06-24T22:34:08Z] FAILED spec_decode/e2e/test_multistep_correctness.py::test_spec_decode_e2e_with_detokenization[1-1-test_llm_kwargs0-per_test_common_llm_kwargs0-common_llm_kwargs0] - ValueError: Memory of devices devices=[0] not free after dur_s=60.06 (threshold_bytes/2**30=2.0)
  | [2024-06-24T22:34:08Z] ===== 10 failed, 284 passed, 2 skipped, 63 warnings in 1002.10s (0:16:42) ======
  | [2024-06-24T22:34:19Z] 🚨 Error: The command exited with status 1
```
I thought that this model might be OK because it is only 3B parameters ~ 12GB in memory in `float32` and it looks like the CI is using a GPU with 24GB memory. 

@cadedaniel Do you know if there are multiple tests running in parallel on the same GPU?

--- a/tests/spec_decode/e2e/test_mlp_correctness.py
+++ b/tests/spec_decode/e2e/test_mlp_correctness.py
@@ -29,7 +29,8 @@ MAIN_MODEL = "ibm-granite/granite-3b-code-instruct"
 # speculative model
 SPEC_MODEL = "ibm-granite/granite-3b-code-instruct-accelerator"
 
-# max. number of speculative tokens
+# max. number of speculative tokens: this corresponds to
+# n_predict in the config.json of the speculator model.
 MAX_SPEC_TOKENS = 5
 
 # precision
@@ -50,17 +51,15 @@ PRECISION = "float16"
 
         # Precision
         "dtype": PRECISION,
-    }])
-@pytest.mark.parametrize("per_test_common_llm_kwargs", [
-    {
+
+        # Main model
         "model": MAIN_MODEL,
-    },
-])
+    }])
+@pytest.mark.parametrize("per_test_common_llm_kwargs", [{}])
 @pytest.mark.parametrize("baseline_llm_kwargs", [{}])
 @pytest.mark.parametrize("test_llm_kwargs", [
     {
         "speculative_model": SPEC_MODEL,
-        "num_speculative_tokens": MAX_SPEC_TOKENS,
     },
 ])
 @pytest.mark.parametrize("output_len", [
@@ -94,17 +93,15 @@ def test_mlp_e2e_greedy_correctness(baseline_llm_generator, test_llm_generator,
 
         # Precision
         "dtype": PRECISION,
-    }])
-@pytest.mark.parametrize("per_test_common_llm_kwargs", [
-    {
+
+        # Main model
         "model": MAIN_MODEL,
-    },
-])
+    }])
+@pytest.mark.parametrize("per_test_common_llm_kwargs", [{}])
 @pytest.mark.parametrize("baseline_llm_kwargs", [{}])
 @pytest.mark.parametrize("test_llm_kwargs", [
     {
         "speculative_model": SPEC_MODEL,
-        "num_speculative_tokens": MAX_SPEC_TOKENS,
     },
 ])
 @pytest.mark.parametrize(
@@ -132,8 +129,6 @@ def test_mlp_e2e_greedy_correctness_with_preemption(baseline_llm_generator,
 @pytest.mark.parametrize(
     "common_llm_kwargs",
     [{
-        "model": MAIN_MODEL,
-
         # Skip cuda graph recording for fast test.
         "enforce_eager": True,
 
@@ -142,6 +137,9 @@ def test_mlp_e2e_greedy_correctness_with_preemption(baseline_llm_generator,
 
         # Precision
         "dtype": PRECISION,
+
+        # Main model
+        "model": MAIN_MODEL,
     }])
 @pytest.mark.parametrize("per_test_common_llm_kwargs", [{}])
 @pytest.mark.parametrize("baseline_llm_kwargs", [{}])
@@ -178,8 +176,6 @@ def test_mlp_different_k(baseline_llm_generator, test_llm_generator,
 @pytest.mark.parametrize(
     "common_llm_kwargs",
     [{
-        "model": MAIN_MODEL,
-
         # Skip cuda graph recording for fast test.
         "enforce_eager": True,
 
@@ -188,13 +184,15 @@ def test_mlp_different_k(baseline_llm_generator, test_llm_generator,
 
         # Precision
         "dtype": PRECISION,
+
+        # Main model
+        "model": MAIN_MODEL,
     }])
 @pytest.mark.parametrize("per_test_common_llm_kwargs", [{}])
 @pytest.mark.parametrize("baseline_llm_kwargs", [{}])
 @pytest.mark.parametrize("test_llm_kwargs",
                          [{
                              "speculative_model": SPEC_MODEL,
-                             "num_speculative_tokens": MAX_SPEC_TOKENS,
                              "speculative_disable_by_batch_size": 4
                          }])
 @pytest.mark.parametrize("batch_size", [1, 5])

--- a/tests/spec_decode/e2e/test_mlp_correctness.py
+++ b/tests/spec_decode/e2e/test_mlp_correctness.py
@@ -33,7 +33,7 @@ SPEC_MODEL = "ibm-granite/granite-3b-code-instruct-accelerator"
 MAX_SPEC_TOKENS = 5
 
 # precision
-PRECISION = "bfloat16"
+PRECISION = "float16"
 
 
 @pytest.mark.parametrize(
@@ -64,7 +64,7 @@ PRECISION = "bfloat16"
     },
 ])
 @pytest.mark.parametrize("output_len", [
-    256,
+    128,
 ])
 @pytest.mark.parametrize("batch_size", [1, 32])
 @pytest.mark.parametrize("seed", [1])
@@ -111,7 +111,7 @@ def test_mlp_e2e_greedy_correctness(baseline_llm_generator, test_llm_generator,
     "output_len",
     [
         # Use small output len for fast test.
-        256,
+        128,
     ])
 @pytest.mark.parametrize("batch_size", [4])
 @pytest.mark.parametrize("seed", [1])

